rbd-mirror A/A: proxy InstanceReplayer APIs via InstanceWatcher RPC

trociny opened the pull request:
Fixes: http://tracker.ceph.com/issues/18787
Signed-off-by: Mykola Golub <mgolub@mirantis.com>

trociny commented:
@dillaman It includes #13803. It might be easier to review only this PR.

dillaman commented:
@trociny OK -- thanks

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -65,6 +68,28 @@ struct RemoveInstanceRequest : public Context {
+  }
+
+  void send() {
+    notifier.notify(bl, nullptr, this);
```
Might be nice to update the API to ```Notifier::notify``` to replace the ```out_bl``` with a pointer to a response struct (map from <gid, cookie> to response payload and list of timed out <gid, cookie> peers). If the response struct has zero peers in the response, you know no instances were registered on the watch.

dillaman commented on src/tools/rbd_mirror/Replayer.cc:
```diff
@@ -801,5 +696,65 @@ void Replayer::handle_pre_release_leader(Context *on_finish) {
+
+  Mutex::Locker locker(m_lock);
+
+  assert(peers.size() == 1U);
```
Nit: I assume this is temporary -- why not directly pass the full collection through?

dillaman commented on src/tools/rbd_mirror/instance_watcher/Types.h:
```diff
@@ -0,0 +1,163 @@
+
+struct ImageAcquirePayload {
+  static const NotifyOp NOTIFY_OP = NOTIFY_OP_IMAGE_ACQUIRE;
+
```
... additionally, if you add a unique request id, you could have a generic "async op complete" callback instead of requiring individual "image acquired" / "image released" messages (similar to ```ImageWatcher```)

trociny commented:
@dillaman I have rebased it against the recent master -- after the pool update notifications have been merged, I don't need ImageMapper to track acquired images, so it is removed from this PR.

I addressed your minor comments, it is still WIP though: need to add notification request id, update the API to Notifier::notify, and address this comment I am not sure I understood:

> Perhaps InstanceWatcher's acquire/release methods should just take a Context * to hide all the retry logic. Then after de-duping requests upon receipt of an acquire/release notification, it can directly invoke the acquire/release methods on InstanceReplayer.

Are you talking about methods like `notify_image_acquire` or `handle_image_acquire`? I suppose the latter?

Do you mean that instead of passing a listener to the constructor, I can pass ImageReplayer (or let InstanceWatcher to create it)? The InstanceWatcher will need also a way to learn who is the leader
currently to send back image_acquired/released notifications. Right now the Replayer uses the LeaderWatcher for this.

dillaman commented:
@trociny I was suggesting that the ```notify_image_acquire``` should take a ```Context *``` so that ```InstanceWatcher``` can hide all the details about retrying failed attempts, ignoring duplicates, etc. The ```InstanceWatcher``` constructor could take a ```InstanceReplayer``` pointer at the time of construction (or at the time of init) so that when it receives an acquire/release notification, it could directly invoke the associated image acquire/release methods within the ```InstanceReplayer``` w/ another ```Context *``` that handles sending the ack -- bypassing the need for a listener.

trociny commented:
@dillaman 
> I was suggesting that the notify_image_acquire should take a Context * so that InstanceWatcher can hide all the details about retrying failed attempts, ignoring duplicates, etc (sort of like the interaction between librbd::Operations and librbd::ImageWatcher). On leader loss, Replayer could invoke a cancel method on InstanceWatcher to remove all tracking for the callbacks so they won't get retried on failure.

Then do we really need image_acquired and image_released notifications? Wouldn't just a successful ack on image_acquire and image_release (on Context * completion) be enough?

> The InstanceWatcher constructor could take a InstanceReplayer pointer at the time of construction (or at the time of init) so that when it receives an acquire/release notification, it could directly invoke the associated image acquire/release methods within the InstanceReplayer w/ another Context * that handles sending the ack -- bypassing the need for a listener.

I could remove "proxy" logic from InstanceWatcher and provide something similar to `Operation::invoke_async_request` in Replayer (or in a separate class/file). The InstanceWatcher notify methods then would just send requests to a specific instance.

> Of course, now that I think about this more, since the ACKs will be sent to the leader's instance watcher, it makes it difficult to wrap all of this logic unless you only ever instatiate a InstanceWatcher for yourself, and the notify_XYZ methods take an instance id to send the messages to the specific instance.

So, instance_id can be provided as a parameter to the InstanceWatcher notify methods (as it is currently), or Policy/ImageMapper could be used by the InstanceWatcher internally to make a decision about the instance. What do you like more?

dillaman commented:
@trociny 

> Then do we really need image_acquired and image_released notifications? Wouldn't just a successful ack on image_acquire and image_release (on Context * completion) be enough?

Assuming we retry on ```-ETIMEDOUT```, consolidate dups, and prevent stale replays (which we should be doing anyway) -- yes, I think we can eliminate the separate ACK messages.

> So, instance_id can be provided as a parameter to the InstanceWatcher notify methods (as it is currently), or Policy/ImageMapper could be used by the InstanceWatcher internally to make a decision about the instance. What do you like more?

I'm fine w/ either -- although since ```InstanceWatcher``` will need additional intelligence, it might make sense for it to eventually interact w/ the policy.

trociny commented:
@dillaman 

> Assuming we retry on -ETIMEDOUT, consolidate dups, and prevent stale replays (which we should be doing anyway) -- yes, I think we can eliminate the separate ACK messages.

Trying this way I am not sure how to "consolidate dups" properly. I assume the source of dups is a situation like below:

- A request is received and being executed by something like: `handle_request(..., on_finish)`, where `on_finish` will eventually complete the watcher's `on_notify_ack`.  
- The handler is executed too long (e.g. stopping a replayer), the sender is timed out and resends the request.
- The dup is received, `handle_request(..., on_finish)` is not executed this time, instead the `on_finish` from the first request is set to use this new `on_notify_ack` on complete.

But I don't know what to do with the first `on_notify_ack`? Just complete with a error code?

There is another situation, when the `handle_request` is completed in a short window after the request is timed out on the sender side, but just before the sender has resent the request. So when receiving a dup, the request is already completed. I think I can handle this properly by keeping the completed request results for some time (2 * TIMEOUT) and just return it in this case.

trociny commented:
@dillaman I have pushed the implementation without image_aquired/image_released ACK messages, for brief review.  I am going to add tests but the moment that worry me right now is handling of "duplicate for in-progress request" in InstanceWatcher::prepare_request(). I am not sure it is correct.

dillaman commented:
> But I don't know what to do with the first on_notify_ack? Just complete with a error code?

You would (under lock) delete it and update your request id -> on_notify_ack mapping to point to the new context. In the case where the leader encountered a -ETIMEOUT error, ack the notification is already too late to accomplish anything. In the case where there was a connection hiccup and the OSD re-sent a duplicate notification to the slave instance, the two different on_notify_ack contexts would essentially be tracking the same request so you can't complete both. 

> There is another situation, when the handle_request is completed in a short window after the request is timed out on the sender side, but just before the sender has resent the request. So when receiving a dup, the request is already completed. I think I can handle this properly by keeping the completed request results for some time (2 * TIMEOUT) and just return it in this case.

In this case, the acquire / release action would have already taken place so you can proceed as normal w/o any additional tracking. Just don't fail if the InstaceReplayer receives a duplicate acquire or release message -- just complete the provided context with a success error code.

trociny commented:
@dillaman 
> You would (under lock) delete it and update your request id -> on_notify_ack mapping to point to the new context. In the case where the leader encountered a -ETIMEOUT error, ack the notification is already too late to accomplish anything. In the case where there was a connection hiccup and the OSD re-sent a duplicate notification to the slave instance, the two different on_notify_ack contexts would essentially be tracking the same request so you can't complete both.

But this looks like we are leaking memory in this case?

> In this case, the acquire / release action would have already taken place so you can proceed as normal w/o any additional tracking. Just don't fail if the InstaceReplayer receives a duplicate acquire or release message -- just complete the provided context with a success error code.

Could you please look at my `InstanceWatcher<I>::prepare_request()` and comment what you see wrong with how it handles dups?

dillaman commented:
@trociny 

> But this looks like we are leaking memory in this case?

Hmm, did valgrind complain about a leak? Assuming you delete the stale ```C_NotifyAck``` context, the low-level librados call just sends a message to the OSD and doesn't deallocate any previously allocated memory slots.

dillaman commented:
@trociny Trying to think about this long-term, the ```Replayer``` should initialize the ```InstanceReplayer``` before ```InstanceWatcher``` and ```LeaderWatcher```. My rationale is that whether or not this instance is the leader, it will be assigned images (by a remote leader or by "itself" if it's the leader).

The ```InstanceWatcher``` could be the proxy for its own notifications if it's provided w/ a reference to its associate ```InstanceReplayer```. That will allow the core changes and separation of ```Replayer``` and ```InstanceReplayer``` to proceed forward w/o supporting RPC to additional peers. Let's get that foundation in place as part of #13803. 

That leaves the question of who is responsible for keeping track of the image assignment state machine given the interaction between ```Replayer::handle_update```, the future image distribution policy class, and ```InstanceWatcher::notify_xyz``` methods. Short term, I think it would be the policy class's responsibility to track the state since handling callbacks requires more knowledge than ```ImageWatcher``` currently has available and I'd like to keep ```Replayer``` small(ish). Until we support additional policies (if ever), I think that is a good approach for reducing the complexity in the short-term.

trociny commented:
@dillaman 
> Trying to think about this long-term, the Replayer should initialize the InstanceReplayer before InstanceWatcher and LeaderWatcher. My rationale is that whether or not this instance is the leader, it will be assigned images (by a remote leader or by "itself" if it's the leader).

Agreed, but right now (before the policies are implemented) if I allow the InstanceReplayer to live when we the Replayer is not the leader, I need a method to tell it to stop all image replayers on the leader release. Just shutting it down does the trick. So I can make the InstanceReplayer long living and provide public `stop_image_replayers` method (which is temporary, as we are not likely to need it later) or do shut down (as I do right now and change this later). Both solutions look like temporal.

> The InstanceWatcher could be the proxy for its own notifications if it's provided w/ a reference to its associate InstanceReplayer.

In my current implementation I am providing InstanceWatcher with the Replayer's callbacks instead of InstanceReplayer mostly because on an image acquire I also need to know the remote io_ctx. I.e. right now the Replayer's callback does the following:
```
  for (auto peer : peers) {
    remote_images.insert({peer.mirror_uuid, peer.image_id, m_remote_io_ctx});
  }
  m_instance_replayer->acquire_image(global_image_id, remote_images, on_finish);
```
So if we want InstanceReplayer is used directly by InstanceWatcher, we need to decide what to do with the remote context (or rather many contexts in future) -- who should be responsible for their initialization and how InstanceWatcher (or just InstanceReplayer) should learn about them.

dillaman commented:
@trociny I think it would be useful to have ```InstanceReplayer``` have a ```stop_all``` command that could be re-used when it determines it has been blacklisted and needs to shut down. That way we can get the current bootstrap ordering for ```InstanceReplayer```.

With regard to the peers, I think you can just add a new ```set_peers``` (or ```add_peer```/```remove_peer``` in the future when we support >1) method to ```InstanceReplayer```. We can clean up the handling of retrieving the remote mirror uuid as well so that it doesn't do a lookup with each ```Replayer::handle_update``` callback.

trociny commented:
@dillaman Ok, I will add `InstanceReplayer::stop_all` though I am not sure we will use it when being blacklisted, because we stop the Replayer in this case, and InstanceReplayer will be shut down.

As for  `InstanceReplayer::set_peers` method, am I correct understanding it this way, that in the Replayer, after initialization of the InstanceReplayer and remote io_ctx, we call `InstanceReplayer::set_peers` (providing remote_mirror_uuid, remote_io_ctx pair(s)), and when later `InstanceReplayer::acquire_image` method is called, its `peers` param will contain remote_mirror_uuid, remote_image_id pair(s), and the InstanceReplayer will be able to match the remote io_ctx by mirror_uuid?

Don't we expect that mirror_uuid can be changed for the pool?

dillaman commented:
@trociny 

> though I am not sure we will use it when being blacklisted, because we stop the Replayer in this case, and InstanceReplayer will be shut down

True -- I am just trying to avoid having to refactor the state machine in the future. I would hope that having ```InstanceReplayer``` iterate through all registered images and invoking the ```release``` method would be a smaller change.

The mirror uuid can only change if (pool) mirroring is disabled and re-enabled on the remote side. I can submit a PR to have ```PoolWatcher``` handle the mirror_uuid retrieval and refresh it if needed.

trociny commented:
@dillaman 
> I can submit a PR to have ```PoolWatcher``` handle the mirror_uuid retrieval and refresh it if needed.

Sure I wouldn't object )

trociny commented:
@dillaman Updated, both this and #13803 (just in case you would like to concentrate on InstanceReplayer only first). Both PR are currently marked DNM, because they depend on #14240.

Also this PR contain DNM hack I am using for testing. It will be removed before merging.

trociny commented:
@dillaman Testing it I found that handling of "duplicate for in-progress request" case in InstanceWatcher<I>::prepare_request was completely wrong.  I updated the code.

trociny commented:
@dillaman It still crashes. I will fix this tomorrow.

trociny commented:
I have fixed the crash. Hope it is much better now.

dillaman commented on src/tools/rbd_mirror/instance_watcher/Types.h:
```diff
@@ -39,8 +39,110 @@ WRITE_CLASS_ENCODER(PeerImageId);
+  NOTIFY_OP_IMAGE_RELEASE  = 1,
+};
+
+struct ImageAcquirePayload {
```
Nit: I would have these two payload classes inherit from a common ```ImagePayloadBase``` class so you don't need to re-implement all the encode/decode/dump methods for the same three parameters.

dillaman commented on src/tools/rbd_mirror/Replayer.h:
```diff
@@ -17,6 +17,7 @@
 #include "include/rados/librados.hpp"
 
 #include "ClusterWatcher.h"
+#include "InstanceWatcher.h"
```
Nit: why can't you forward declare anymore?

dillaman commented on src/tools/rbd_mirror/LeaderWatcher.h:
```diff
@@ -46,6 +46,8 @@ class LeaderWatcher : protected librbd::Watcher {
   void shut_down(Context *on_finish);
 
   bool is_leader();
+  bool is_releasing_leader();
```
Nit: can these methods be const?

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.h:
```diff
@@ -4,54 +4,74 @@
-                  const boost::optional<std::string> &id = boost::none);
+  InstanceWatcher(Threads<ImageCtxT> *threads, librados::IoCtx &io_ctx,
+                  InstanceReplayer<ImageCtxT> *instance_replayer,
+                  const boost::optional<std::string> &id);
```
Nit: can we remove the optional ```id```?

trociny replied:
Do you mean to remove `id` argument, or to make it non-optional?

The `id` is mostly needed only for `InstanceWatcher::remove_instance()` (and may be for tests). So it is not for public use (and `create` method does not have it).

I think I can make it non-optional in the constructor and properly set in `create()`.

dillaman replied:
OK, sounds good

dillaman commented on src/tools/rbd_mirror/instance_watcher/Types.h:
```diff
@@ -39,8 +39,110 @@ WRITE_CLASS_ENCODER(PeerImageId);
+
+  uint64_t request_id;
+  std::string global_image_id;
+  PeerImageIds peers;
```
I wonder if we can eliminate this from the acquire/release notifications. Each rbd-mirror daemon instance (currently) tracks its peers (albeit only a single peer right now). The leader instance doesn't know which peer (if any) are primary, so regardless it's up to the assigned instance's ImageReplayer to figure that out. Therefore, on acquire the peer instance should have enough information to start. On the release side, just because an image disappears doesn't necessarily imply the primary was deleted. Again, right now we have ```ImageDeleter``` checking to see if we should really remove the image. I am thinking the ```ImageDeleter``` hooks that currently exist within ```InstanceReplayer``` should be moved to ```ImageReplayer```. I'll try to work on this idea tomorrow in the current codebase.

trociny replied:
The source of information about peer is the PoolWatcher, which detects a mirroring image addition/removal in the pool. I thought pool watchers would be run only by the instance leader?

Now I think it is more correct to have only one peer in acquire/release notifications (not a list), because a message is going to be generated on a pool state change. Then actually we might not need `PeerImageId` type (and PeerImage). I added those mainly to be able to pass a list of peers (remote images). When it is always one peer, we can just pass mirror_uuid, image_id as arguments. And the newly introduced ImageReplayer::set_remote_images() should be removed and only add_remote_image()/remove_remote_image() are used.

dillaman replied:
The peers actually come from "Mirror" when it scans the rbd_mirroring object and creates one Replayer per peer (eventually needs to be on Replayer per pool w/ multiple peers). Since an image will eventually be available on all peers, it doesn't really matter which specific peer has the image (or had the image removed) so long as we get notifications to re-verify a given image so that we can determine if we should remove it locally or if we should register ourselves in the journal of the peer.

trociny replied:
Sorry, by peer (in the `PoolWatcher` context) I meant `ImagePeer`.

I am still not sure I see how we can nicely eliminate `ImagePeer` from the acquire/release notifications.

Let's we have two peers (remote mirroring pools). An image (primary) is created in one of the pools. The `PoolWatcher` sends update to the `Replayer`, and the `Replayer` sends image_acquire notification. At this moment the `Replayer` possesses information about the image peer (remote image) and can provide it within `image_acquire` notification. If it is not sent to the `InstanceReplayer`, it will need to figure this out itself before calling `ImageReplayer::add_remote_image`, i.e. scan all available peers for an image with this global id. Do you want me to do this way?

dillaman replied:
When we eventually support multiple peers, ```ImageReplayer``` will need to scan all peers regardless. Take the example where the ```PoolWatcher``` gets a image added update from peer C -- but peer C's image was actually just a non-primary replica of peer B's primary image.

trociny replied:
If it is `ImageReplayer` who should scan all peers, then I suppose `ImageReplayer::add_remote_image/remove_remote_image` methods should retire too? You added them recently so I thought you did not plan this would be the `ImageReplayer` job. Then we need to add  `ImageReplayer::add_peer/remove_peer` instead?

Talking about the race you described, it does not look like hurt and will resolve eventually. Anyway, the `ImageReplayer` looks like will need to recover from such situations to choose a proper peer image. Still the images themselves could be provided externally.

I am ok to do this either way, just want to make it clear.

dillaman replied:
Agreed -- definitely won't hurt. I'm just trying to say that passing around a collection of peers on each image acquire/release probably doesn't matter. If we pass only the single peer that was associated with the original event, perhaps that could provide some optimization in the future so perhaps it doesn't hurt to replace the collection w/ a single peer. If we don't end up using it, at least it can be useful for debugging events.

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -49,24 +56,117 @@ struct RemoveInstanceRequest : public Context {
   }
 };
 
+struct NotifyInstanceRequest : public Context {
```
Nit: usually prefix these w/ ```C_```

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -165,12 +269,64 @@ void InstanceWatcher<I>::remove(Context *on_finish) {
-  acknowledge_notify(notify_id, handle, out);
+  Mutex::Locker locker(m_lock);
+
+  if (m_on_finish != nullptr) {
```
Leaks ```on_notify_ack```

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -165,12 +269,64 @@ void InstanceWatcher<I>::remove(Context *on_finish) {
+
+  Mutex::Locker locker(m_lock);
+
+  if (m_on_finish != nullptr) {
```
Nit: same comment here

trociny commented:
@dillaman updated

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -166,12 +272,57 @@ void InstanceWatcher<I>::remove(Context *on_finish) {
+    uint64_t request_id = ++m_request_seq;
+    bufferlist bl;
+    ::encode(NotifyMessage{ImageAcquirePayload{request_id, global_image_id,
+        peer_mirror_uuid, peer_image_id}}, bl);
```
Nit: spacing

trociny replied:
@dillaman not sure what indent you like in this case. `peer_mirror_uuid` should start from the same position as `request_id` above?

dillaman replied:
@trociny yeah, preferably it should be indented to the same level as its associated paramters (```request_id``` in this case) or move all the parameters down to a new (indented) line.

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -166,12 +272,57 @@ void InstanceWatcher<I>::remove(Context *on_finish) {
+    uint64_t request_id = ++m_request_seq;
+    bufferlist bl;
+    ::encode(NotifyMessage{ImageReleasePayload{request_id, global_image_id,
+        peer_mirror_uuid, peer_image_id, schedule_delete}}, bl);
```
Nit: spacing

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -87,18 +185,26 @@ void InstanceWatcher<I>::remove_instance(librados::IoCtx &io_ctx,
 
 template <typename I>
+InstanceWatcher<I> *InstanceWatcher<I>::create(
+  librados::IoCtx &io_ctx, ContextWQ *work_queue,
```
Nit: 4 spaces indent is cleaner for a hanging indent

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -166,12 +272,57 @@ void InstanceWatcher<I>::remove(Context *on_finish) {
-                                       uint64_t notifier_id, bufferlist &bl) {
-  dout(20) << dendl;
+void InstanceWatcher<I>::notify_image_acquire(
+  const std::string &instance_id, const std::string &global_image_id,
```
Nit: 4 spaces indent is cleaner for a hanging indent

dillaman commented on src/tools/rbd_mirror/InstanceWatcher.cc:
```diff
@@ -46,28 +51,121 @@ struct C_GetInstances : public Context {
+        if (r == -ETIMEDOUT) {
+          derr << "C_NotifyInstanceRequest: " << this << " " << __func__
+               << ": resending after timeout" << dendl;
+          send();
```
We are going to need some way to cancel in-flight notifications (e.g. leader died so we will never get a response from it). All in-flight notifications would be canceled on shut down and notifications to a specific peer would be canceled when the leader blacklists a peer due to missed heartbeats.

trociny commented:
@dillaman updated:

- indent issues fixed
- cancel for C_NotifyInstanceRequest added
- TestMockInstanceWatcher.ImageAcquireRelease improved (which helped to find a bug in C_NotifyInstanceRequest)
- TestMockInstanceWatcher.mageAcquireReleaseCancel added

I have not added a hook for canceling in-flight notifications to an instance that is being removed. It looks like not difficult to do by passing InstanceWatcher to Instances, but I would rather to work on  it later as a separate PR. The same problem will be with cancelling SYNC notifications, and we might need something more generic.
